compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1620,6 +1620,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
     def findEnclosingThis(moduleClass: Symbol, from: Symbol): Type =
       if ((from.owner eq moduleClass) && from.isPackageObject && from.is(Opaque)) from.thisType
+      else if (from eq moduleClass.owner) && moduleClass.isTopLevelDefinitionsObject && moduleClass.is(Opaque) then moduleClass.thisType
```

Comment:
Isn't that going to create a `C$.this` reference *outside* of `C$`? If yes, I don't think that's supposed to be possible/valid.

Reply:
The reference wasn't outside when the user wrote it.  The compiler moved it so it's nested in a package object, but to the user the call site isn't from "outside".

Reply:
I understand, but still ... from an elaborated, TASTy-level point of view, this is invalid. A `C.this` reference has *no meaning* outside of `C`. This is worse than being "not accessible" because of `protected` or even "not visible" because of `private`; here it does not even *exist*.

Reply:
I'd argue that C.this, where C is a singleton module is just an alternative to C.type.  Which, btw, is what `widenInferred` does as a final cleanup.

Reply:
> I'd argue that C.this, where C is a singleton module is just an alternative to C.type.

It represents the same set of values at run-time, sure, but it's not valid everywhere `C.type` is valid.

> Which, btw, is what `widenInferred` does as a final cleanup.

AFAICT, `widentInferred` turns `pre.C$` into `pre.C`; not `C$.this` into `pre.C` nor the converse.